### PR TITLE
Sort imports in Phobos

### PR DESCRIPTION
**Pull request by wilzbach**
This is a preparation of the upcoming [Dscanner enforcement for sorted imports](https://github.com/dlang-community/D-Scanner/blob/master/src/analysis/imports_sortedness.d) in Phobos.
Of course, this wasn't [done by hand](https://github.com/wilzbach/dscanner-fix/blob/master/sort_imports.d).

There are still a couple of false negative to filter out at the Dscanner plugin (e.g. https://github.com/dlang-community/D-Scanner/pull/444), but this import sorting script was laying on my disk now for a couple of months and we need to sort the imports anyways before we can move ahead with enforcing sortedness.

CC @andralex @CyberShadow 

As this is quite a big chunk, I would appreciate if I don't need to rebase this that often ;-)

### CODE REVIEW

File: std/algorithm/comparison.d [hunk 1]
```diff
@@ -925,7 +925,7 @@ range of range (of range...) comparisons.
 
 @safe pure unittest
 {
-    import std.utf : byChar, byWchar, byDchar;
+    import std.utf : byChar, byDchar, byWchar;
```
Comment (wilzbach) [inline std/algorithm/comparison.d]: @andralex if we want to enforce sortedness for all imports, we have to look over a few minor exceptions where the order makes a bit of sense.

File: std/datetime/systime.d [hunk 2]
```diff
@@ -1758,7 +1758,7 @@ public:
     ///
     @safe unittest
     {
-        import core.time : msecs, usecs, hnsecs, nsecs;
+        import core.time : hnsecs, msecs, nsecs, usecs;
```
Comment (wilzbach) [inline std/datetime/systime.d]: That's the other example I found where order made sense.

**Comment by jmdavis** [comment]
Sure, it makes sense to be sorting imports alphabetically in general, but does it really make sense to effectively make it an error when they're not sorted that way? That seems like it's getting really picky about something that doesn't really matter much, especially if most imports are local, and we have fewer, large blocks of import statements. And of course, forcing sorting causes problems in the cases where another order arguably makes more sense.

**Comment by andralex** [comment]
It would be odd to have the compiler enforce it, but in phobos it's reasonable. The problem with other orders is that everybody may have their own notion of the best order.

File: std/algorithm/comparison.d (hunk 1 above)
Comment (andralex) [inline std/algorithm/comparison.d]: I'm not too worried about the few hand-optimizations that could be done. I'd say sorted is the simple and effective way.

**Comment by andralex** [comment]
Thanks. Should I override jenkins?

**Comment by andralex** [comment]
Apparently I can... we really should get jenkins going.

**Comment by CyberShadow** [comment]
I'm really not sure about this one (for the reasons @jmdavis mentioned).

**Comment by CyberShadow** [comment]
I think this should be reverted.

> The problem with other orders is that everybody may have their own notion of the best order.

This is a slippery slope towards having one single "normalized" form of the code, and deviating from it is punishable by a 10-minute edit/amend/push/wait-for-CI cycle, and if we continue moving in this direction, I can see it happening a lot. Having code follow a consistent style, but this is going beyond style - many of the changes make the code uglier simply for the sake of normalization. If we are to enforce that imports are sorted, then why not also enforce that local variables are also sorted? Why should it be allowed to write `int minValue, maxValue;` if `int maxValue, minValue;` is a more "normal" form?

**Comment by wilzbach** [comment]
> This is a slippery slope towards having one single "normalized" form of the code, and deviating from it is punishable by a 10-minute edit/amend/push/wait-for-CI cycle,

FWIW You can run dscanner locally - it's one of the most popular D projects & some IDEs/editors even have native support.

> Having code follow a consistent style is good, but this is going beyond style - many of the changes make the code uglier simply for the sake of normalization.

I only found two examples where order actually made sense.
The entire point of the normalization is that we don't need to spend time anymore on nitpicking the order of imports.

**Comment by CyberShadow** [comment]
> You can run dscanner locally

This is going to be completely irrelevant to most submitters unless they are told how to do it during the submission process. Links hidden among walls of text don't count. Even then, it lessens the frustration only to some degree.

**Comment by wilzbach** [comment]
> This is going to be completely irrelevant to most submitters unless they are told how to do it during the submission process. Links hidden among walls of text don't count. Even then, it lessens the frustration only to some degree.

This is getting OT a bit, but FWIW we do plan to show a [hello word message](https://github.com/dlang-bots/dlang-bot/pull/44) to new contributors. Also will this message contain detailed instructions based on the errors, see e.g. https://github.com/dlang-bots/dlang-bot/pull/57

**Comment by andralex** [comment]
>I think this should be reverted.

This causes no harm regardless of how we move forward, so please leave it in. Thanks.

>This is a slippery slope towards having one single "normalized" form of the code, and deviating from it is punishable by a 10-minute edit/amend/push/wait-for-CI cycle, and if we continue moving in this direction, I can see it happening a lot.

Yah, if we go for the tooling approach we should make it trivial for people to make the check locally - perhaps even automatically as part of the make process.

A variety of successful projects have very strict rules for formatting. I know from a GNU tools maintainer that their "good diff" tool (I forgot the name) was a lifesaver.

File: std/algorithm/comparison.d (hunk 1 above)
Comment (CyberShadow) [inline std/algorithm/comparison.d]: :-1:

File: std/complex.d [hunk 3]
```diff
@@ -370,7 +370,7 @@ if (isFloatingPoint!T)
     ref Complex opOpAssign(string op, C)(C z)
         if (op == "^^" && is(C R == Complex!R))
     {
-        import std.math : exp, log, cos, sin;
+        import std.math : cos, exp, log, sin;
```
Comment (CyberShadow) [inline std/complex.d]: :-1:

File: std/complex.d [hunk 4]
```diff
@@ -298,7 +298,7 @@ if (isFloatingPoint!T)
     Complex!(CommonType!(T, R)) opBinaryRight(string op, R)(R lhs) const
         if (op == "^^" && isNumeric!R)
     {
-        import std.math : cos, exp, log, sin, PI;
+        import std.math : cos, exp, log, PI, sin;
```
Comment (CyberShadow) [inline std/complex.d]: :-1:

File: std/complex.d [hunk 5]
```diff
@@ -799,7 +799,7 @@ Complex!T conj(T)(Complex!T z) @safe pure nothrow @nogc
 Complex!(CommonType!(T, U)) fromPolar(T, U)(T modulus, U argument)
     @safe pure nothrow @nogc
 {
-    import std.math : sin, cos;
+    import std.math : cos, sin;
```
Comment (CyberShadow) [inline std/complex.d]: :-1:

File: std/container/array.d [hunk 6]
```diff
@@ -251,7 +251,7 @@ private struct RangeT(A)
 struct Array(T)
 if (!is(Unqual!T == bool))
 {
-    import core.stdc.stdlib : malloc, realloc, free;
+    import core.stdc.stdlib : free, malloc, realloc;
```
Comment (CyberShadow) [inline std/container/array.d]: :-1:

File: std/container/dlist.d [hunk 7]
```diff
@@ -39,7 +39,7 @@ module std.container.dlist;
 
     // If you want to apply range operations, simply slice it.
     import std.algorithm.searching : countUntil;
-    import std.range : popFrontN, popBackN, walkLength;
+    import std.range : popBackN, popFrontN, walkLength;
```
Comment (CyberShadow) [inline std/container/dlist.d]: :-1:

File: std/conv.d [hunk 8]
```diff
@@ -1851,7 +1851,7 @@ if (isInputRange!S && !isInfinite!S && isSomeChar!(ElementEncodingType!S) &&
 // bugzilla 15800
 @safe unittest
 {
-    import std.utf : byCodeUnit, byChar, byWchar, byDchar;
+    import std.utf : byChar, byCodeUnit, byDchar, byWchar;
```
Comment (CyberShadow) [inline std/conv.d]: :-1:

File: std/datetime/systime.d (hunk 2 above)
Comment (CyberShadow) [inline std/datetime/systime.d]: :-1:

File: std/datetime/systime.d [hunk 9]
```diff
@@ -1846,7 +1846,7 @@ public:
     ///
     @safe unittest
     {
-        import core.time : Duration, msecs, hnsecs, nsecs;
+        import core.time : Duration, hnsecs, msecs, nsecs;
```
Comment (CyberShadow) [inline std/datetime/systime.d]: :-1:

File: std/datetime/systime.d [hunk 10]
```diff
@@ -4564,7 +4564,7 @@ public:
     ///
     @safe unittest
     {
-        import core.time : msecs, hnsecs;
+        import core.time : hnsecs, msecs;
```
Comment (CyberShadow) [inline std/datetime/systime.d]: :-1:

File: std/datetime/systime.d [hunk 11]
```diff
@@ -7383,7 +7383,7 @@ public:
     ///
     @safe unittest
     {
-        import core.time : msecs, usecs, hnsecs;
+        import core.time : hnsecs, msecs, usecs;
```
Comment (CyberShadow) [inline std/datetime/systime.d]: :-1:

File: std/datetime/systime.d [hunk 12]
```diff
@@ -8310,7 +8310,7 @@ public:
     ///
     @safe unittest
     {
-        import core.time : hours, msecs, usecs, hnsecs;
+        import core.time : hnsecs, hours, msecs, usecs;
```
Comment (CyberShadow) [inline std/datetime/systime.d]: :-1:

File: std/datetime/systime.d [hunk 13]
```diff
@@ -8559,7 +8559,7 @@ public:
...
```
Comment (CyberShadow) [inline std/datetime/systime.d]: :-1:

File: std/datetime/systime.d [hunk 14]
```diff
@@ -8785,7 +8785,7 @@ public:
...
```
Comment (CyberShadow) [inline std/datetime/systime.d]: :-1:

File: std/algorithm/iteration.d [hunk 15]
```diff
@@ -3219,7 +3219,7 @@ if (fun.length >= 1)
     // Sum all elements with explicit seed
     assert(arr.fold!((a, b) => a + b)(6) == 21);
 
-    import std.algorithm.comparison : min, max;
+    import std.algorithm.comparison : max, min;
```
Comment (CyberShadow) [inline std/algorithm/iteration.d]: :-1:

File: std/algorithm/sorting.d [hunk 16]
```diff
@@ -2038,8 +2038,8 @@ if (((ss == SwapStrategy.unstable && (hasSwappableElements!Range ||
 
 private void quickSortImpl(alias less, Range)(Range r, size_t depth)
 {
+    import std.algorithm.comparison : max, min;
     import std.algorithm.mutation : swap, swapAt;
-    import std.algorithm.comparison : min, max;
```
Comment (CyberShadow) [inline std/algorithm/sorting.d]: :-1:

File: std/array.d [hunk 17]
```diff
@@ -744,7 +744,7 @@ slice, returns that slice. Otherwise, returns the null slice.
 auto overlap(T, U)(T[] r1, U[] r2) @trusted pure nothrow
 if (is(typeof(r1.ptr < r2.ptr) == bool))
 {
-    import std.algorithm.comparison : min, max;
+    import std.algorithm.comparison : max, min;
```
Comment (CyberShadow) [inline std/array.d]: :-1:

File: std/experimental/allocator/building_blocks/free_tree.d [hunk 18]
```diff
@@ -52,7 +52,7 @@ struct FreeTree(ParentAllocator)
     static assert(ParentAllocator.alignment % size_t.alignof == 0,
         "FreeTree must be on top of a word-aligned allocator");
 
-    import std.algorithm.comparison : min, max;
+    import std.algorithm.comparison : max, min;
```
Comment (CyberShadow) [inline std/experimental/allocator/building_blocks/free_tree.d]: :-1:

File: std/experimental/allocator/package.d [hunk 19]
```diff
@@ -229,15 +229,15 @@ public import std.experimental.allocator.common,
 // Example in the synopsis above
 @system unittest
 {
-    import std.algorithm.comparison : min, max;
-    import std.experimental.allocator.building_blocks.free_list : FreeList;
-    import std.experimental.allocator.gc_allocator : GCAllocator;
-    import std.experimental.allocator.building_blocks.segregator : Segregator;
-    import std.experimental.allocator.building_blocks.bucketizer : Bucketizer;
+    import std.algorithm.comparison : max, min;
```
Comment (CyberShadow) [inline std/experimental/allocator/package.d]: :-1:

**Comment by wilzbach** [comment]
Btw @CyberShadow we have had sortedness of imports as part of the [DStyle for Phobos](http://dlang.org/dstyle.html) since some time now.

> Yah, if we go for the tooling approach we should make it trivial for people to make the check locally - perhaps even automatically as part of the make process.

`style_lint` is already a Makefile target - adding it to `unittest` (or a more generic `test`) would be rather easy ...

> A variety of successful projects have very strict rules for formatting. I know from a GNU tools maintainer that their "good diff" tool (I forgot the name) was a lifesaver.

... and we don't swear at people using the wrong comment style:
http://lkml.iu.edu/hypermail/linux/kernel/1607.1/00627.html

File: std/getopt.d [hunk 20]
```diff
@@ -1614,8 +1614,8 @@ Params:
 */
 void defaultGetoptFormatter(Output)(Output output, string text, Option[] opt)
 {
+    import std.algorithm.comparison : max, min;
     import std.format : formattedWrite;
-    import std.algorithm.comparison : min, max;
```
Comment (CyberShadow) [inline std/getopt.d]: :-1:

File: std/parallelism.d [hunk 21]
```diff
@@ -3931,12 +3931,12 @@ version(unittest)
 // These are the tests that should be run every time Phobos is compiled.
 @system unittest
 {
+    import std.algorithm.comparison : equal, max, min;
     import std.algorithm.iteration : filter, map, reduce;
-    import std.algorithm.comparison : equal, min, max;
```
Comment (CyberShadow) [inline std/parallelism.d]: :-1:

File: std/range/package.d [hunk 22]
```diff
@@ -4197,7 +4197,7 @@ if (Ranges.length && allSatisfy!(isInputRange, Ranges))
                     return ranges[0].length;
 
                 //[min|max](ranges[0].length, ranges[1].length, ...)
-                import std.algorithm.comparison : min, max;
+                import std.algorithm.comparison : max, min;
```
Comment (CyberShadow) [inline std/range/package.d]: :-1:

File: std/array.d [hunk 23]
```diff
@@ -76,13 +76,13 @@ Source: $(PHOBOSSRC std/_array.d)
...
+static import std.algorithm.iteration; // FIXME, remove with alias of splitter
+import std.functional;
 import std.meta;
 import std.traits;
-import std.functional;
-static import std.algorithm.iteration; // FIXME, remove with alias of splitter
 
 import std.range.primitives;
-public import std.range.primitives : save, empty, popFront, popBack, front, back;
+public import std.range.primitives : back, empty, front, popBack, popFront, save;
```
Comment (CyberShadow) [inline std/array.d]: :-1:

File: std/numeric.d [hunk 24]
```diff
@@ -128,7 +128,7 @@ if (((flags & flags.signed) + precision + exponentWidth) % 8 == 0 && precision +
 ///
 @safe unittest
 {
-    import std.math : sin, cos;
+    import std.math : cos, sin;
```
Comment (CyberShadow) [inline std/numeric.d]: :-1:

File: std/digest/digest.d [hunk 25]
```diff
@@ -1136,7 +1136,7 @@ if (isInputRange!R1 && isInputRange!R2 && !isInfinite!R1 && !isInfinite!R2 &&
     import std.internal.test.dummyrange : ReferenceInputRange;
     import std.range : takeExactly;
     import std.string : representation;
-    import std.utf : byWchar, byDchar;
+    import std.utf : byDchar, byWchar;
```
Comment (CyberShadow) [inline std/digest/digest.d]: :-1:

File: std/path.d [hunk 26]
```diff
@@ -641,7 +641,7 @@ if (isRandomAccessRange!R && hasSlicing!R && hasLength!R && isSomeChar!(ElementT
     static assert(dirName("dir/file") == "dir");
 
     import std.array;
-    import std.utf : byChar, byWchar, byDchar;
+    import std.utf : byChar, byDchar, byWchar;
```
Comment (CyberShadow) [inline std/path.d]: :-1:

File: std/path.d [hunk 27]
```diff
@@ -1075,7 +1075,7 @@ if (isConvertibleToString!R)
     assert(stripExtension("file.ext1.ext2"d) == "file.ext1");
 
     import std.array;
-    import std.utf : byChar, byWchar, byDchar;
+    import std.utf : byChar, byDchar, byWchar;
```
Comment (CyberShadow) [inline std/path.d]: :-1:

File: std/string.d [hunk 28]
```diff
@@ -579,7 +579,7 @@ if (isConvertibleToString!Range)
     import std.conv : to;
     import std.exception : assertCTFEable;
     import std.traits : EnumMembers;
-    import std.utf : byChar, byWchar, byDchar;
+    import std.utf : byChar, byDchar, byWchar;
```
Comment (CyberShadow) [inline std/string.d]: :-1:

File: std/string.d [hunk 29]
```diff
@@ -3396,7 +3396,7 @@ if (isConvertibleToString!Range)
 
     // Ranges
     import std.array : array;
-    import std.utf : byChar, byWchar, byDchar;
+    import std.utf : byChar, byDchar, byWchar;
```
Comment (CyberShadow) [inline std/string.d]: :-1:

File: std/string.d [hunk 30]
```diff
@@ -3496,7 +3496,7 @@ unittest
...
```
Comment (CyberShadow) [inline std/string.d]: :-1:

File: std/string.d [hunk 31]
```diff
@@ -3608,7 +3608,7 @@ if (isConvertibleToString!Range)
 @safe pure unittest
 {
     import std.array : array;
-    import std.utf : byChar, byWchar, byDchar, byCodeUnit, invalidUTFstrings;
+    import std.utf : byChar, byCodeUnit, byDchar, byWchar, invalidUTFstrings;
```
Comment (CyberShadow) [inline std/string.d]: :-1:

File: std/string.d [hunk 32]
```diff
@@ -6545,7 +6545,7 @@ if ((isInputRange!Range && isSomeChar!(Unqual!(ElementEncodingType!Range)) ||
 ///
 @safe pure unittest
 {
-    import std.utf : byChar, byWchar, byDchar;
+    import std.utf : byChar, byDchar, byWchar;
```
Comment (CyberShadow) [inline std/string.d]: :-1:

File: std/experimental/checkedint.d [hunk 33]
```diff
@@ -2277,7 +2277,7 @@ if (isIntegral!L && isIntegral!R)
     else
         alias Result = typeof(mixin("L() " ~ x ~ " R()"));
 
-    import core.checkedint : addu, adds, subs, muls, subu, mulu;
+    import core.checkedint : adds, addu, muls, mulu, subs, subu;
```
Comment (CyberShadow) [inline std/experimental/checkedint.d]: :-1:

File: std/numeric.d [hunk 34]
```diff
@@ -2316,7 +2316,7 @@ optimizations.
 struct GapWeightedSimilarityIncremental(Range, F = double)
 if (isRandomAccessRange!(Range) && hasLength!(Range))
 {
-    import core.stdc.stdlib : malloc, realloc, alloca, free;
+    import core.stdc.stdlib : alloca, free, malloc, realloc;
```
Comment (CyberShadow) [inline std/numeric.d]: :-1:

File: std/numeric.d [hunk 35]
```diff
@@ -2187,10 +2187,10 @@ F gapWeightedSimilarity(alias comp = "a == b", R1, R2, F)(R1 s, R2 t, F lambda)
 if (isRandomAccessRange!(R1) && hasLength!(R1) &&
     isRandomAccessRange!(R2) && hasLength!(R2))
 {
-    import std.functional : binaryFun;
-    import std.algorithm.mutation : swap;
-    import core.stdc.stdlib : malloc, free;
     import core.exception : onOutOfMemoryError;
+    import core.stdc.stdlib : free, malloc;
```
Comment (CyberShadow) [inline std/numeric.d]: :-1:

File: std/parallelism.d [hunk 36]
```diff
@@ -2625,7 +2625,7 @@ public:
             byte[maxStack] buf = void;
             immutable size_t nBytesNeeded = nWorkUnits * RTask.sizeof;
 
-            import core.stdc.stdlib : malloc, free;
+            import core.stdc.stdlib : free, malloc;
```
Comment (CyberShadow) [inline std/parallelism.d]: :-1:

File: std/parallelism.d [hunk 37]
```diff
@@ -3372,7 +3372,7 @@ private void submitAndExecute(
     immutable nThreads = pool.size + 1;
 
     alias PTask = typeof(scopedTask(doIt));
-    import core.stdc.stdlib : malloc, free;
+    import core.stdc.stdlib : free, malloc;
```
Comment (CyberShadow) [inline std/parallelism.d]: :-1:

File: std/path.d [hunk 38]
```diff
@@ -3849,9 +3849,9 @@ string expandTilde(string inputPath) nothrow
 {
     version(Posix)
     {
-        import core.stdc.stdlib : malloc, free, realloc;
         import core.exception : onOutOfMemoryError;
-        import core.stdc.errno : errno, ERANGE;
+        import core.stdc.errno : ERANGE, errno;
+        import core.stdc.stdlib : free, malloc, realloc;
```
Comment (CyberShadow) [inline std/path.d]: :-1:

File: std/regex/package.d [hunk 39]
```diff
@@ -692,7 +692,7 @@ public:
 if (isSomeString!R)
 {
 private:
-    import core.stdc.stdlib : malloc, free;
+    import core.stdc.stdlib : free, malloc;
```
Comment (CyberShadow) [inline std/regex/package.d]: :-1:

File: std/regex/package.d [hunk 40]
```diff
@@ -810,7 +810,7 @@ public:
 
 private @trusted auto matchOnce(alias Engine, RegEx, R)(R input, RegEx re)
 {
-    import core.stdc.stdlib : malloc, free;
+    import core.stdc.stdlib : free, malloc;
```
Comment (CyberShadow) [inline std/regex/package.d]: :-1:

File: std/algorithm/mutation.d [hunk 41]
```diff
@@ -901,7 +901,7 @@ if (is(Range == char[]) || is(Range == wchar[]))
 ///
 @system unittest
 {
-    import core.stdc.stdlib : malloc, free;
+    import core.stdc.stdlib : free, malloc;
```
Comment (CyberShadow) [inline std/algorithm/mutation.d]: :-1:

File: std/algorithm/mutation.d [hunk 42]
```diff
@@ -2835,7 +2835,7 @@ if (isInputRange!Range && hasLvalueElements!Range && is(typeof(range.front = val
 ///
 nothrow @system unittest
 {
-    import core.stdc.stdlib : malloc, free;
+    import core.stdc.stdlib : free, malloc;
```
Comment (CyberShadow) [inline std/algorithm/mutation.d]: :-1:

File: std/process.d [hunk 43]
```diff
@@ -386,7 +386,7 @@ private Pid spawnProcessImpl(in char[][] args,
     scope(exit) if (workDirFD >= 0) close(workDirFD);
     if (workDir.length)
     {
-        import core.sys.posix.fcntl : open, O_RDONLY, stat_t, fstat, S_ISDIR;
+        import core.sys.posix.fcntl : fstat, open, O_RDONLY, stat_t, S_ISDIR;
```
Comment (CyberShadow) [inline std/process.d]: :-1:

File: std/process.d [hunk 44]
```diff
@@ -778,7 +778,7 @@ version (Posix) @safe unittest
 version (Posix)
 private void setCLOEXEC(int fd, bool on) nothrow @nogc
 {
-    import core.sys.posix.fcntl : fcntl, F_GETFD, FD_CLOEXEC, F_SETFD;
+    import core.sys.posix.fcntl : fcntl, FD_CLOEXEC, F_GETFD, F_SETFD;
```
Comment (CyberShadow) [inline std/process.d]: :+1:

File: std/string.d [hunk 45]
```diff
@@ -635,7 +635,7 @@ if (isConvertibleToString!Range)
 {
     import std.conv : to;
     import std.traits : EnumMembers;
-    import std.utf : byCodeUnit, byChar, byWchar;
+    import std.utf : byChar, byCodeUnit, byWchar;
```
Comment (CyberShadow) [inline std/string.d]: :-1:

File: std/windows/registry.d [hunk 46]
```diff
@@ -82,7 +82,7 @@ class Win32Exception : WindowsException
     @property int error() { return super.code; }
 }
 
-version(unittest) import std.string : startsWith, endsWith;
+version(unittest) import std.string : endsWith, startsWith;
```
Comment (CyberShadow) [inline std/windows/registry.d]: :-1:

File: std/string.d [hunk 47]
```diff
@@ -199,8 +199,8 @@ import std.traits; // isConvertibleToString, isNarrowString, isSomeChar,
 
 //public imports for backward compatibility
 public import std.algorithm.comparison : cmp;
-public import std.algorithm.searching : startsWith, endsWith, count;
-public import std.array : join, replace, replaceInPlace, split, empty;
+public import std.algorithm.searching : count, endsWith, startsWith;
```
Comment (CyberShadow) [inline std/string.d]: :-1:

File: std/windows/syserror.d [hunk 48]
```diff
@@ -180,9 +180,9 @@ T wenforce(T)(T condition, const(char)[] name, const(wchar)* namez, string file
 version(Windows)
 @system unittest
 {
+    import std.algorithm.searching : endsWith, startsWith;
     import std.exception;
     import std.string;
-    import std.algorithm.searching : startsWith, endsWith;
```
Comment (CyberShadow) [inline std/windows/syserror.d]: :-1:

**Comment by CyberShadow** [comment]
> This causes no harm regardless of how we move forward,

Not true, this is a regression in code readability. Many, many small regressions, but a step backwards nevertheless.

> A variety of successful projects have very strict rules for formatting.

Formatting is one thing; normalization at the cost of readability is another. Which is it?

> we have had sortedness of imports as part of the DStyle for Phobos since some time now.

The style is ambiguous on whether the rule applies to selective imports as well.

**Comment by jmdavis** [comment]
> This is a slippery slope towards having one single "normalized" form of the code, and deviating from it is punishable by a 10-minute edit/amend/push/wait-for-CI cycle, and if we continue moving in this direction, I can see it happening a lot. 

It happens to me almost every time I create a pull request. From my perspective, the CI stuff has largely made life far more annoying for minimal benefit.

> Yah, if we go for the tooling approach we should make it trivial for people to make the check locally - perhaps even automatically as part of the make process.

That would improve things in some ways and make them worse in others. I didn't even know that the 'style' target existed until Seb pointed out it to me at dconf, but even knowing about it, I can't use it on my Ubuntu system, because it doesn't work with `PIC=1`. I've have enough trouble building the docs historically, because of all of the extra stray tools that the dlang.org makefiles want, forcing me to edit the makefile just to be able to build it without all of that extra junk so that I can see what the docs look like. Making phobos depend on other tools is just going to make things worse. Now, if all of the parts required are just D things that get built as part of the build, then fine, but I don't want to have to constantly fight the druntime and Phobos builds, because they're overly complicated in terms of what they require.

It seems to me that all of this extra stuff is making it harder to contribute, not easier.

**Comment by andralex** [comment]
>That would improve things in some ways and make them worse in others. I didn't even know that the 'style' target existed until Seb pointed out it to me at dconf, but even knowing about it, I can't use it on my Ubuntu system, because it doesn't work with PIC=1.

Most definitely a smooth operation is essential for the success of tooling. Style checks 

>It seems to me that all of this extra stuff is making it harder to contribute, not easier.

There's a lot of value in having automated tools instead of reviewers point out simple things like brace placement, punctuations, indentation etc. It's much more effective for both contributors and reviewers.

**Comment by jmdavis** [comment]
> There's a lot of value in having automated tools instead of reviewers point out simple things like brace placement, punctuations, indentation etc. It's much more effective for both contributors and reviewers.

Oh, it can definitely help, but simply not being picky about something makes it even easier. Some level of pickiness on style makes sense so that the code is reasonably consistent, but it really feels to me like it's gone too far. Even with proper tooling that points out problems locally before creating a PR, complaining about something like the fact that `startsWith` being imported before `endsWith` just creates extra work for the person creating the PR without really improving the code.

Regardless, as it stands, the end result is that many of us create PRs only to have the Circle CI scream at us over some nitpicky thing, and then we have to go through several cycles to make it happy.

**Comment by andralex** [comment]
@jmdavis most certainly the hodge-podge of styles present in Phobos right now does not make it easier for anyone involved.

@wilzbach how about we start with only sorting imports alphabetically, not symbols within imports. The original problem was that people missed the presence of an import and would add it again.

**Comment by CyberShadow** [comment]
> @wilzbach how about we start with only sorting imports alphabetically, not symbols within imports.

https://github.com/dlang/phobos/pull/5481

**Pull request merged**